crypto/src/fixpoint.rs

```diff
@@ -36,17 +39,47 @@ impl Debug for U128x128 {
     }
 }
 
-// This is a hack to make the `Display` impl work.
-fn f64_from_secret(value: &U128x128) -> f64 {
+/// Converts an `f64` into a `U128x128`, this method will panic and is lossy.
+/// It should only be used in low stakes client-side contexts.
+/// Chesterton's documentation: `panic!` are to discourage usage of this method.
+pub fn from_f64_unsafe(value: f64) -> U128x128 {
+    if value < 0.0 {
+        panic!("U128x128 cannot represent negative numbers");
+    }
+
+    if value.is_infinite() {
+        panic!("U128x128 cannot represent infinite values");
+    }
+
+    if value.is_nan() {
+        panic!("U128x128 cannot represent NaN values");
+    }
+
+    // Split the f64 into its integral and fractional parts.
+    let integral = value as u128;
+    let fractional = value.fract();
+
+    // Convert the fractional part to an unsigned 128-bit integer.
+    // 1. Multiply the fractional part by 2^128 to move the significant bits to the left.
+    // 2. Round down the result to the nearest integer.
+    // 3. Convert the rounded result to an unsigned 128-bit integer.
+    let fractional_as_u128 = (fractional * f64::from_bits(BASE)).trunc() as u128;
+
+    // Combine the integral and fractional parts into a U128x128.
+    let combined = U256::from_words(integral, fractional_as_u128);
+    U128x128(combined)
+}
```

Comment:
I think this should probably just be a `TryFrom` impl, and we should document that it's only an approximate conversion. It definitely shouldn't panic, though, I don't think that's a good way to discourage bad use, it just makes everything more crashy.

Reply:
Fair enough, this is fixed in the latest batch of changes!